### PR TITLE
Added Service for Auto Discovery of Feeds for URLs

### LINKED ISSUE #1645: Create Service for Auto discovery of Feeds for URLs
**What would you like to be added**:

Add a way for an RSS/Atom Feed to be discovered for a URL.  This requires a few things:

1. back-end API that takes a URL and examines/tests it to determine a Feed URL.  For example, if I give you http://blog.com I might get back http://blog.com/feed.  We could do this as a microservice as part of #1627.
2. front-end component(s) that allow a user to enter a URL, trigger an API call to our feed discovery back-end service, then show the resulting feed URL or an error.

Most blogging platforms follow a pattern we can define in code:

- Wordpress: http://blog-user-name.wordpress.com/category/category-name/feed/
- Blogger: http://blog-user-name.blogspot.com/feeds/posts/default/-/categorylabel?alt=rss
- Medium: https://medium.com/feed/@blog-user-name
- Dev.to: https://dev.to/feed/blog-user-name

There are others that we can probably guess or brute force to discover.  We can also do some validation in the back-end to make sure a feed URL is valid (returns data we can parse).

**Why would you like this to be added**:

#1643 is going aiming to improve the sign-up experience for users, and this is one aspect of that work.

Reference: 
Link to current feed list: https://wiki.cdot.senecacollege.ca/wiki/Planet_CDOT_Feed_List

### PR DESCRIPTION
**Pull request by tonyvugithub**
## Issue This PR Addresses
Fixed #1645 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
Followed the layout of image microservice to add the feed discovery service. The service will return a list of potential feed urls if the user provided a valid blog url

## Checklist

image

Unit Tests
image
image

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by tonyvugithub** [comment]
Right now for all the well-known blog provider this works perfectly. Some edge case like personal website blog like https://alexwang.ca/feed the service cannot discover the feed url.

**Comment by chrispinkney** [comment]
Hello! What do you think about adding some tests? These tests can ensure that the service works in case changes ever need to be made, or more likely, the blog provider changes exactly how an RSS feed is created created on their end.

File: src/api/feedDiscovery/src/lib/discover.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    // Extract the list of atom link elements
+    const atom =
+      $('link[type="application/atom+xml"]').clone() ||
+      $('link[type="application/x.atom+xml"]').clone() ||
+      $('link[type="application/x-atom+xml"]').clone();
+    const feedList = rss.length ? rss : atom; // If rss has any content (i.e. it actually exists), assign its value to elem, otherwise assign the value of atom
+    const feedUrl = feedList[0].attribs.href;
+    return feedUrl;
+  } catch (e) {
+    throw new Error();
```
Comment (chrispinkney) [inline src/api/feedDiscovery/src/lib/discover.js:29]: Should `e` be fed into `Error();`?

**Review by chrispinkney — COMMENTED** [review]
Also the `.gitignore` is empty and can be removed, unless you think we need it for something

File: src/api/feedDiscovery/src/lib/discover.js (hunk 1 above)
Comment (tonyvugithub) [reply src/api/feedDiscovery/src/lib/discover.js:29]: I still don't know what would be best to handling error. This is a work in progress.

**Comment by tonyvugithub** [comment]
@chrispinkney : Yeah I will start writing tests soon. Just want to have some review on the functionality of it first.

File: src/api/feedDiscovery/src/lib/discover.js (hunk 1 above)
Comment (chrispinkney) [reply src/api/feedDiscovery/src/lib/discover.js:29]: Curious about this too, my service does the same thing. Would love to know if a better way for this exists.

Comment (humphd) [reply src/api/feedDiscovery/src/lib/discover.js:29]: In general, it's an anti-pattern to do this:

```js
try {
  ...
} catch(err) {
  throw new Error();
}
```

You just lost all context from the original error that could have been logged or used to decide what to do next.  If you're going simply re-throw the error, remove the try/catch (i.e., it's not doing anything) and let the caller deal with the exception.

Should we validate that this/these feed URLs are working before we continue?

File: src/api/feedDiscovery/src/lib/discover.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+// A middleware to check if the URL is a valid web page, 200 status with text/html type
+async function checkValidWebPage(req, res, next) {
+  try {
+    const { statusCode, headers } = await got.head(req.body.blogUrl);
+    const contentType = headers['content-type'];
+    // If status code is not 200 or content-type is not text/html then send 400 error
+    if (statusCode !== 200 || !contentType.includes('text/html')) {
+      res.status(400).send();
```
Comment (humphd) [inline src/api/feedDiscovery/src/lib/discover.js:50]: Create a proper HTTP error, like this: https://github.com/Seneca-CDOT/satellite/blob/master/index.js#L113

File: src/api/feedDiscovery/src/lib/discover.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  next();
+}
+
+// A middleware to check if the URL is a valid web page, 200 status with text/html type
+async function checkValidWebPage(req, res, next) {
+  try {
+    const { statusCode, headers } = await got.head(req.body.blogUrl);
+    const contentType = headers['content-type'];
+    // If status code is not 200 or content-type is not text/html then send 400 error
+    if (statusCode !== 200 || !contentType.includes('text/html')) {
```
Comment (humphd) [inline src/api/feedDiscovery/src/lib/discover.js:49]: `if(!(statusCode === 200 && contentType.includes('text/html')))`

File: src/api/feedDiscovery/src/lib/discover.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    res.status(400).send('Invalid Blog URL Provided');
+  }
+  // Else, continue
+  next();
+}
+
+// A middleware to check if the URL is a valid web page, 200 status with text/html type
+async function checkValidWebPage(req, res, next) {
+  try {
+    const { statusCode, headers } = await got.head(req.body.blogUrl);
```
Comment (humphd) [inline src/api/feedDiscovery/src/lib/discover.js:46]: Do a `HEAD` request so we don't download the whole thing

File: src/api/feedDiscovery/src/lib/discover.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    const contentType = headers['content-type'];
+    // If status code is not 200 or content-type is not text/html then send 400 error
+    if (statusCode !== 200 || !contentType.includes('text/html')) {
+      res.status(400).send();
+    } else {
+      next();
+    }
+  } catch (err) {
+    // if there is err (eg: 404 status), return 400 error
+    res.status(400).send();
```
Comment (humphd) [inline src/api/feedDiscovery/src/lib/discover.js:56]: Proper HTTP error here

File: src/api/feedDiscovery/src/routes/feedDiscover.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+const { Router } = require('@senecacdot/satellite');
+const { checkValidUrl, checkValidWebPage, discoverFeedUrl } = require('../lib/discover');
+
+const router = Router();
+
+router.post('/', [checkValidUrl, checkValidWebPage, discoverFeedUrl], (req, res) => {
+  res.status(200).send({
+    feedUrl: res.locals.feedUrl,
```
Comment (humphd) [inline src/api/feedDiscovery/src/routes/feedDiscover.js:8]: `urls: res.locals.urls`

**Review by humphd — CHANGES_REQUESTED** [review]
I did a first pass through this and left some feedback.

**Comment by humphd** [comment]
Out of interest I wrote a script to parse all ~6500 blog posts I have in Redis and most have more than 1 `<link rel="alternate">` element, often like this:

```
  'application/rss+xml',
  'application/json',
  'application/json+oembed',
  'text/xml+oembed'
```

Consider a post like this one: https://neilong31.wordpress.com/2019/12/07/release-0-4-the-end/

It has 5:

```
application/rss+xml https://neilong31.wordpress.com/feed/
application/rss+xml https://neilong31.wordpress.com/comments/feed/
application/rss+xml https://neilong31.wordpress.com/2019/12/07/release-0-4-the-end/feed/
application/json+oembed https://public-api.wordpress.com/oembed/?format=json&url=htt…%2F12%2F07%2Frelease-0-4-the-end%2F&for=wpcom-auto-discovery
application/xml+oembed https://public-api.wordpress.com/oembed/?format=xml&url=http…%2F12%2F07%2Frelease-0-4-the-end%2F&for=wpcom-auto-discovery
```

So I think we probably need to return an array of potential feed URLs, and the UI can offer suggestions for the user to pick from (the first is maybe the right one, maybe not).

File: src/api/feedDiscovery/src/lib/discover.js (hunk 1 above)
Comment (tonyvugithub) [reply src/api/feedDiscovery/src/lib/discover.js:29]: @humphd : I was thinking about it but then I realized that those feed urls will be extracted from the provider, would that mean it's already legit? So I skip it.

**Review by humphd — CHANGES_REQUESTED** [review]
Let's take this out of Draft if you're actually working on it.

File: src/api/feed-discovery/src/lib/discover.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  const $ = cheerio.load(res.body);
+  // Extract the list of rss link elements
+  const rss = $('link[type="application/rss+xml"]').clone();
+  // Extract the list of atom link elements
+  const atom =
+    $('link[type="application/atom+xml"]').clone() ||
+    $('link[type="application/x.atom+xml"]').clone() ||
+    $('link[type="application/x-atom+xml"]').clone();
+  // If rss has any content (i.e. it actually exists), assign its value to elem, otherwise assign the value of atom
+  const feedList = rss.length ? rss : atom;
```
[the remark this answers is not in the document]
Comment (tonyvugithub) [reply src/api/feed-discovery/src/lib/discover.js:19]: I was thinking to give priority to rss so if there is rss feed then work with that. If not then go with Atom. Should I return all of them? Also is there an example of atom feed url for testing purpose? Most of the link I checked they have rss feed.

[the remark this answers is not in the document]
Comment (humphd) [reply src/api/feed-discovery/src/lib/discover.js:19]: Your service is validating and discovering feeds, not picking them.  Return all options.

We have thousands of examples: https://wiki.cdot.senecacollege.ca/wiki/Planet_CDOT_Feed_List#Feeds

File: src/api/feed-discovery/server.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+const { Satellite } = require('@senecacdot/satellite');
+
+const feedDiscoverRouter = require('./src/routes/feed-discover');
```
[the remark this answers is not in the document]
Comment (tonyvugithub) [reply src/api/feed-discovery/server.js:3]: @humphd : I tried to follow the same pattern you did for the image service where you have the router files inside the `routes` directory. I guess it does not matter?

**Comment by tonyvugithub** [comment]
> Out of interest I wrote a script to parse all ~6500 blog posts I have in Redis and most have more than 1 `<link rel="alternate">` element, often like this:
> 
> ```
>   'application/rss+xml',
>   'application/json',
>   'application/json+oembed',
>   'text/xml+oembed'
> ```
> 
> Consider a post like this one: https://neilong31.wordpress.com/2019/12/07/release-0-4-the-end/
> 
> It has 5:
> 
> ```
> application/rss+xml https://neilong31.wordpress.com/feed/
> application/rss+xml https://neilong31.wordpress.com/comments/feed/
> application/rss+xml https://neilong31.wordpress.com/2019/12/07/release-0-4-the-end/feed/
> application/json+oembed https://public-api.wordpress.com/oembed/?format=json&url=htt…%2F12%2F07%2Frelease-0-4-the-end%2F&for=wpcom-auto-discovery
> application/xml+oembed https://public-api.wordpress.com/oembed/?format=xml&url=http…%2F12%2F07%2Frelease-0-4-the-end%2F&for=wpcom-auto-discovery
> ```
> 
> So I think we probably need to return an array of potential feed URLs, and the UI can offer suggestions for the user to pick from (the first is maybe the right one, maybe not).

@humphd : Right now in my service for rss I only considered `application/rss+xml`, should I consider the other alternate links as you mentioned in comment above?

File: src/api/feedDiscovery/src/lib/discover.js (hunk 1 above)
Comment (humphd) [reply src/api/feedDiscovery/src/lib/discover.js]: We could try trusting them for now, sure.

File: src/api/feed-discovery/server.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (humphd) [reply src/api/feed-discovery/server.js:3]: The image service has two routers, you only have one, so let's simplify.

**Comment by humphd** [comment]
> @humphd : Right now in my service for rss I only considered `application/rss+xml`, should I consider the other alternate links as you mentioned in comment above?

Yes, absolutely.

File: src/api/feed-discovery/server.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+const service = new Satellite();
+
+service.router.use('/', feedDiscoverRouter);
+
+const port = parseInt(process.env.FEED_DISCOVERY_PORT || 5555, 10);
+
+service.start(port, () => {
+  console.log(`Service is running on port ${port}`);
```
Comment (humphd) [inline src/api/feed-discovery/server.js:12]: Get rid of this console.log.

File: src/api/feed-discovery/src/lib/discover.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
+/* eslint-disable no-plusplus */
+/* eslint-disable consistent-return */
```
Comment (humphd) [inline src/api/feed-discovery/src/lib/discover.js:2]: Let's fix the problems in the code vs. disabling eslint if possible.

File: src/api/feed-discovery/src/lib/discover.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    return next(createError(400, 'Invalid Blog URL'));
+  }
+  // Else, continue
+  next();
+}
+
+// A middleware to check if the URL is a valid web page, 200 status with text/html type
+async function checkValidBlogPage(req, res, next) {
+  try {
+    const { statusCode, headers } = await got.head(req.body.blogUrl);
```
Comment (humphd) [inline src/api/feed-discovery/src/lib/discover.js:56]: It looks like we're going to have to download the HEAD and BODY of the blog in order to fulfill this route's needs.  Instead of doing it twice, why don't you download the full response (GET vs. HEAD) and store the body on locals to pass to the next middleware?

**Review by humphd — CHANGES_REQUESTED** [review]
This needs tests too.

File: src/api/feed-discovery/src/lib/discover.js (hunk 11 above)
Comment (tonyvugithub) [reply src/api/feed-discovery/src/lib/discover.js:56]: That works too, Let me update it!

File: src/api/feed-discovery/src/lib.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    return next(createError(400, 'Invalid Blog URL'));
+  }
+  // Else, continue
+  return next();
+}
+
+// A middleware to check if the URL is a valid web page, 200 status with text/html type
+async function checkValidBlogPage(req, res, next) {
+  try {
+    const { statusCode, headers } = await got.head(req.body.blogUrl);
```
Comment (humphd) [inline src/api/feed-discovery/src/lib.js:54]: Why `.head`?  Can't we get the full thing and pass it along to the next middleware?

**Review by humphd — CHANGES_REQUESTED** [review]
Your `.gitignore` is empty.  Remove the file if you don't need it.

What's your plan for testing?

**Comment by tonyvugithub** [comment]
@humphd : I am thinking to do some unit tests to test out the middlewares. Have not got a chance to look at your e2e tests for image service deeply. But I think I will do some of that too. My plan was to have a working version of it and then write the tests.

File: src/api/feed-discovery/src/lib.js (hunk 12 above)
Comment (tonyvugithub) [reply src/api/feed-discovery/src/lib.js:54]: I am updating it in next commit. Probably later today

**Comment by humphd** [comment]
You won't need e2e tests for this.  You can do mock network requests to simulate all kinds of cases: blog URLs that don't work, ones that do, ones that have an RSS feed, ones that have RSS and Atom, etc.  You can make sure that you get back the expected result each time.  It should be pretty quick once you figure out how to mock the network requests to return the HTML/RSS/ATOM content you expect (we have other tests that do this now).

File: src/api/feed-discovery/server.js [hunk 13]
```diff
@@ -0,0 +1,11 @@
+const { Satellite } = require('@senecacdot/satellite');
+
+const feedDiscoverRouter = require('./src/router');
+
+const service = new Satellite();
```
Comment (humphd) [inline src/api/feed-discovery/server.js:5]: You might want to split this file in two to make testing easier, see my code in https://github.com/Seneca-CDOT/telescope/pull/1893/files where I have `src/index.js` and `src/server.js` to make getting at the service's `app` instance easier.

File: src/api/feed-discovery/src/router.js [hunk 14]
```diff
@@ -0,0 +1,12 @@
+const { Router } = require('@senecacdot/satellite');
+const { checkValidUrl, checkValidBlogPage, discoverFeedUrls } = require('./lib');
+
+const router = Router();
+
+router.post('/', [checkValidUrl, checkValidBlogPage, discoverFeedUrls], (req, res) => {
+  res.status(200).send({
```
Comment (humphd) [inline src/api/feed-discovery/src/router.js:7]: `.json({...})`

**Review by humphd — CHANGES_REQUESTED** [review]
Looks good, a few small things.

File: src/api/feed-discovery/src/lib.js [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+const cheerio = require('cheerio');
+const validUrl = require('valid-url');
+const createError = require('http-errors');
+
+// Helper function to return the feed url of a given blog url
+function getFeedUrls(document) {
+  // Load the html document
+  const $ = cheerio.load(document);
+  // Extract the Cheerio object of rss link elements
+  const rss = $('link[type="application/rss+xml"]').clone();
```
Comment (humphd) [inline src/api/feed-discovery/src/lib.js:11]: Question: why do you `.clone()` in all cases, and not just use the object returned?  Also, there's a lot of code duplication in this function.

I think we should simplify a bit.  For starters, we could combine how you're collecting all these links, since each one is just a similar selector:

```js
const selectors = [
  'link[type="application/rss+xml"]',
  'link[type="application/atom+xml"]',
  ...
].join(',');
const links = $(selectors);
```
Then you can do your iteration once:

```js
const feedUrls = [];

if(links) {
  links.each((i, elem) => {
    const url = elem.attr('href');
    if(href) {
      feedUrls.push(url);
    }
  });
}
```

**Review by humphd — CHANGES_REQUESTED** [review]
Looking good, one simplification I think you can make.  NOTE: my code is untested, so you may have to adapt it.

File: src/api/feed-discovery/src/lib.js (hunk 15 above)
Comment (tonyvugithub) [reply src/api/feed-discovery/src/lib.js]: @humphd : done! Thanks for the feedbacks. I am still learning how Cheerio works so this has been at the back of my mind.

**Comment by humphd** [comment]
Can you not squash new commits?  It makes it harder to review, since I have to re-read the whole thing.  Better to push fixes to review comments and squash at the end.

**Review by humphd — COMMENTED** [review]
This looks good.  Now it needs tests to prove it works.

Let me know if you need help with those, and I can guide you.

**Comment by tonyvugithub** [comment]
@humphd : I keep in mind for future commits. I have been doing this for a while. I guess for issue with many files, it's not optimal. Sorry for any inconvenience.

File: src/api/feed-discovery/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,14 @@
+FROM node:lts-alpine as base
+
+# https://snyk.io/blog/10-best-practices-to-containerize-nodejs-web-applications-with-docker/
+RUN apk add dumb-init
+
+WORKDIR /app
+
+COPY --chown=node:node . .
+
+RUN npm install ci --only=production
```
Comment (humphd) [inline src/api/feed-discovery/Dockerfile:10]: Remove the `ci`

**Review by humphd — CHANGES_REQUESTED** [review]
I'll review fully later, but that screenshot is caused by this.

**Review by chrispinkney — APPROVED** [review]
Postman_2021-03-12_18-01-15

Nice work man. Just tested this via Docker and it LGTM. Would love to see what others think too.

Do you have any intention of adding some sort of logging (via Satellite's `Logger`) somewhere? Might help if an issue goes wrong. Might not be necessary though since you're at least returning 400 and an error message.

Postman_2021-03-12_18-09-19

File: src/api/feed-discovery/test/router.test.js [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+      const expectedError = createError(400, 'Invalid Blog URL');
+
+      // Expect the next() function to be called with the expected error
+      expect(mockNextFunction).toBeCalledWith(expectedError);
+
+      request(app)
+        .post('/')
+        .send({ blogUrl: invalidUrl })
+        // Expect to receive 400 status and expected response body
+        .expect(400, '<h1>400 Error</h1><p>Invalid Blog URL</p>', done);
```
Comment (humphd) [inline src/api/feed-discovery/test/router.test.js:31]: Let's not depend on the actual HTML that gets returned, just the 400.  We might change it later.

File: src/api/feed-discovery/test/router.test.js [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+        'Content-Type': 'text/xml',
+      });
+
+      request(app).post('/').send({ blogUrl: InvalidBlogPage }).expect(400, done);
+    });
+  });
+
+  describe('Test discoverFeedUrls middleware', () => {
+    it('should return 404 if there is no feed url discovered', (done) => {
+      const validUrl = 'https://LinkWithNoFeedUrls.com';
```
Comment (humphd) [inline src/api/feed-discovery/test/router.test.js:71]: all lowercase URLs please

File: src/api/feed-discovery/test/router.test.js [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+  });
+
+  it('should return 200 status and if all middlewares got passed', async (done) => {
+    const blogUrl = 'https://test321.blogspot.com/';
+    const mockBlogUrlResponseBody = `
+      <!doctype html>
+      <html lang="en">
+        <head>
+          <link rel="alternate" type="application/atom+xml" href="https://test321.blogspot.com/feeds/posts/default/-/open-source"/>
+          <link rel="alternate" type="application/rss+xml" href="https://test321.blogspot.com/feeds/posts/default/-/open-source?alt=rss"/>
```
Comment (humphd) [inline src/api/feed-discovery/test/router.test.js:130]: This is a good start, but add tests for every feed type that you search for in that Array.  Make sure they all work.

**Review by humphd — CHANGES_REQUESTED** [review]
Nice.  A few more tests are needed, but it's close.

**Comment by tonyvugithub** [comment]
@chrispinkney : I don't think it's necessary since the return json is self explanatory if something went wrong

**Review by chrispinkney — DISMISSED** [review]

File: src/api/feed-discovery/test/router.test.js [hunk 20]
```diff
@@ -93,7 +93,7 @@ describe('POST /', () => {
     });
 
     it('should call next() if there is any discovered feed url', (done) => {
-      const validUrl = 'https://LinkWithFeedUrls.com';
+      const validUrl = 'https://LinkWithFeedURLs.com';
```
Comment (humphd) [inline src/api/feed-discovery/test/router.test.js:96]: URLs should *always* be lowercase.  You'll drive yourself crazy with debugging on systems that are case sensitive vs. not, otherwise. 

```js
const validUrl = 'https://linkwithfeedurls.com';
```

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/feed-discovery/test/router.test.js [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+      const mockNextFunction = jest.fn();
+      checkValidUrl(mockReq, mockRes, mockNextFunction);
+      expect(mockNextFunction).toHaveBeenCalledTimes(1);
+      done();
+    });
+  });
+
+  describe('Test checkValidBlogPage middleware', () => {
+    it('should return 400 if the blog page responses with code other than 200', (done) => {
+      const InvalidBlogPage = 'https://notExistBlogPage.com/';
```
Comment (humphd) [inline src/api/feed-discovery/test/router.test.js:51]: All lower case: `nonexistingblogpage.com`

File: src/api/feed-discovery/test/router.test.js [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+
+      nock(InvalidBlogPage).get('/').reply(404);
+
+      request(app).post('/').send({ blogUrl: InvalidBlogPage }).expect(400, done);
+    });
+
+    it('should return 400 if the blog page responses with content type other than text/html', (done) => {
+      const InvalidBlogPage = 'https://notHtmlResponse.com/';
```
Comment (humphd) [inline src/api/feed-discovery/test/router.test.js:59]: `https://nonhtmlresponse.com`

File: src/api/feed-discovery/test/router.test.js (hunk 18 above)
Comment (humphd) [reply src/api/feed-discovery/test/router.test.js:71]: This wasn't fixed

File: src/api/feed-discovery/test/router.test.js (hunk 20 above)
Comment (tonyvugithub) [reply src/api/feed-discovery/test/router.test.js:96]: Ah, i see. I misunderstood your point. Will update shortly.

File: src/api/feed-discovery/test/router.test.js [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+    });
+  });
+
+  it('should return 200 status and if all middlewares got passed', async (done) => {
+    const blogUrl = 'https://test321.blogspot.com/';
+    const mockBlogUrlResponseBody = `
+      <!doctype html>
+      <html lang="en">
+        <head>
+          <link rel="alternate" type="application/atom+xml" href="https://test321.blogspot.com/feeds/posts/default/-/open-source"/>
```
Comment (humphd) [inline src/api/feed-discovery/test/router.test.js:129]: This is OK, but a tip: you're testing too many things at once.  A test should only test a single thing.  For example, you can test having no feed URLs (you have that), one RSS (you have that), one Atom, multiple feeds, tests for various types separated into different files.

If you do a giant test that fails, it's hard to debug: which part of the test is failing?  If you separate these out into smaller tests that only test 1 thing, it's way easier to figure out what's broken.

**Review by humphd — CHANGES_REQUESTED** [review]

**Review by humphd — DISMISSED** [review]
Well done.

File: src/api/feed-discovery/test/lib.test.js [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+      const mockRes = {};
+      const mockNextFunction = jest.fn();
+      checkValidUrl(mockReq, mockRes, mockNextFunction);
+      expect(mockNextFunction).toHaveBeenCalledTimes(1);
+      done();
+    });
+  });
+
+  describe('Test checkValidBlogPage middleware', () => {
+    it('should return 400 if the blog page responses with code other than 200', (done) => {
```
Comment (humphd) [inline src/api/feed-discovery/test/lib.test.js:50]: responds vs. responses

File: src/api/feed-discovery/test/lib.test.js [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+      const InvalidBlogPage = 'https://notexistblogpage.com/';
```
Comment (humphd) [inline src/api/feed-discovery/test/lib.test.js:51]: invalidBlogPage (lowercase i)

File: src/api/feed-discovery/test/lib.test.js [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+
+      nock(InvalidBlogPage).get('/').reply(404);
+
+      request(app).post('/').send({ blogUrl: InvalidBlogPage }).expect(400, done);
+    });
+
+    it('should return 400 if the blog page responses with content type other than text/html', (done) => {
+      const InvalidBlogPage = 'https://nothtmlresponse.com/';
```
Comment (humphd) [inline src/api/feed-discovery/test/lib.test.js:59]: Same here, and throughout: lowerCamelCase for JS naming.

**Review by humphd — CHANGES_REQUESTED** [review]
A few small style things, and this is ready.  So cool to have this.  Can't wait to hook it up to the sign in flow.

**Review by chrispinkney — APPROVED** [review]

**Review by humphd — APPROVED** [review]
